fabric-particles-v1/src/main/java/net/fabricmc/fabric/mixin/client/particles/MixinParticleManager.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.client.particles;
+
+import java.util.List;
+import java.util.Map;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Accessor;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import net.fabricmc.fabric.impl.particles.FabricParticleManager;
+import net.fabricmc.fabric.impl.particles.VanillaParticleManager;
+import net.minecraft.client.particle.ParticleFactory;
+import net.minecraft.client.particle.ParticleManager;
+import net.minecraft.client.particle.ParticleTextureData;
+import net.minecraft.client.texture.SpriteAtlasTexture;
+import net.minecraft.resource.ResourceManager;
+import net.minecraft.util.Identifier;
+
+@Mixin(ParticleManager.class)
+public abstract class MixinParticleManager implements VanillaParticleManager {
+
+    private final FabricParticleManager fabricParticleManager = new FabricParticleManager(this);
+
+    @Override
+    @Accessor("particleAtlasTexture")
+    public abstract SpriteAtlasTexture getAtlas();
+
+    @Override
+    @Accessor("factories")
+    public abstract Int2ObjectMap<ParticleFactory<?>> getFactories();
+
+    @Inject(method = "registerDefaultFactories()V", at = @At("RETURN"))
+    private void onRegisterDefaultFactories(CallbackInfo info) {
+        fabricParticleManager.injectValues();
+    }
+
+    @Redirect(method = "method_18836(Lnet/minecraft/client/resource/ResourceManager;Lnet/minecraft/util/Identifier;Ljava/util/Map;)V",
```

Comment:
This should probably not be a redirect, since it may create compatibility-related crashes (not as badly as an overwrite, but still enough to be a concern). If reasonably possible, I'd replace it with an injection instead.

Reply:
This is a bit tricky. For this to work I need to have a reference to `ParticleTextureData` (a local) _and_ the parameters to the parent method.

I can do a injectino here, but I'd be missing the local, or I can inject into `ParticleTextureData` but I'd be missing the contextual values from here.

Redirect conveniently seems to be the only method that provides both.

Reply:
Relating to this, I have a possible solution to get rid of the redirect. However I'm tired and can't test it right now, so I'll get back to this tomorrow and push it if everything still works.

Reply:
Tested and pushed! It now uses an Inject at HEAD and if our system knows of a particle with a matching ID, we do the de-serializing and cancel the upstream method.